### ISSUE TITLE
Access to XMLHttpRequest at 'https://bam.nr-data.net/1/undefined?...' from origin 'xxxx' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.

### ISSUE THREAD

**Issue by php4fan** [newrelic/newrelic-browser-agent#950]
[NOTE]: # ( ^^ Provide a general summary of the issue in the title above. ^^ )

## Description

I hope this is the right repository.
Sometimes in some pages of some apps, NewRelic's client-side script throws this error.
There have been reports of this in the forum forever, but I haven't been able to find any issue report here.

## Steps to Reproduce

Visit any of these urls:
https://bio-parques.com/el-mapa-de-los-bioparques-de-espana-t92.html
https://forokarting.com/el-mapa-de-circuitos-de-karting-de-espana-t84.html
https://www.foro-bomberos.com/el-mapa-de-los-parques-de-bomberos-de-espana-t2997.html

Look at the console

## Expected Behavior

There should be no javascript errors in the console, or at least none thrown by NewRelic's script.
Also, I'm pretty sure the URL of the script injected by NewRelic shouldn't have the path `/1/undefined`

## Observed Behavior

This error shows up in the console:
```
Access to XMLHttpRequest at 'https://bam.nr-data.net/1/undefined?a=undefined&v=1.255.0&t=Unnamed%20Transaction&rst=1653&ck=0&s=7a504df49a0d5bcd&ref=https://bio-parques.com/el-mapa-de-los-bioparques-de-espana-t92.html&hr=0&be=348&fe=1242&dc=658&fsh=1&perf=%7B%22timing%22:%7B%22of%22:1712484928654,%22n%22:0,%22f%22:1,%22dn%22:54,%22dne%22:54,%22c%22:54,%22s%22:80,%22ce%22:132,%22rq%22:132,%22rp%22:348,%22rpe%22:428,%22di%22:996,%22ds%22:996,%22de%22:1006,%22dc%22:1550,%22l%22:1550,%22le%22:1590%7D,%22navigation%22:%7B%7D%7D&fp=974&fcp=974' from origin 'https://bio-parques.com' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```

Also note multiple occurrences of the substring `"undefined"` in the URL, especially the one in the very URL path. I wonder if that is the core of the issue.

## Relevant Logs / Console output

```
Access to XMLHttpRequest at 'https://bam.nr-data.net/1/undefined?a=undefined&v=1.255.0&t=Unnamed%20Transaction&rst=1653&ck=0&s=7a504df49a0d5bcd&ref=https://bio-parques.com/el-mapa-de-los-bioparques-de-espana-t92.html&hr=0&be=348&fe=1242&dc=658&fsh=1&perf=%7B%22timing%22:%7B%22of%22:1712484928654,%22n%22:0,%22f%22:1,%22dn%22:54,%22dne%22:54,%22c%22:54,%22s%22:80,%22ce%22:132,%22rq%22:132,%22rp%22:348,%22rpe%22:428,%22di%22:996,%22ds%22:996,%22de%22:1006,%22dc%22:1550,%22l%22:1550,%22le%22:1590%7D,%22navigation%22:%7B%7D%7D&fp=974&fcp=974' from origin 'https://bio-parques.com' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```

## Your Environment

* Server: PHP 7.4 on Debian, latest PHP agent
* Client: Chrome on Manjaro Linux

## Additional context

[TIP]:  # ( Add any other context about the problem here. )

**Comment by patrickhousley** [comment]
Hi @php4fan the error you provided is an indication that you have not configured the agent correctly. Looking at the first link, it appears you are missing key configuration properties. Please review our [installation docs](https://docs.newrelic.com/docs/browser/browser-monitoring/installation/install-browser-monitoring-agent/) and reach out to the New Relic support team at https://support.newrelic.com/s/ for additional assistance.

**Comment by php4fan** [comment]
> Looking at the first link, it appears you are missing key configuration properties

If that was really the case, I should get a warning about that when I enter one.newrelic.com.
It is unacceptable to get console errors at the user's end and have to figure out what keys I am supposedly missing in my configuration.

Also, that doesn't seem to make any sense. All configuration properties should have defaults. I can't think of anything that can't have a default besides the license key, which is definitely not missing, and even if there was something else, then the browser code injected into the client should simply be empty, rather than produce a script that triggers errors (and again, a gigantic warning should pop up somewhere at the newrelic dashboard).

No matter how wrong my configuration is (and if it is, it's what was given to me by the installer), the behavior is not acceptable. This should be reopened and looked at more seriously.

**Comment by patrickhousley** [comment]
Hi @php4fan the github repository issues is mainly for help with the NPM package. We have a whole team of support engineers dedicated to helping with issues with the agent. I would suggest reaching out to them [here](https://support.newrelic.com/s/). There is a link on that page, once you login, to contact support where you can open a ticket.

**Comment by php4fan** [comment]
> We have a whole team of support engineers dedicated to helping with issues with the agent. I would suggest reaching out to them [here](https://support.newrelic.com/s/).

Yeah, I've been waiting for them to reply for 4 days
https://forum.newrelic.com/s/hubtopic/aAXPh0000001FwHOAU/javascript-injected-by-newrelic-throws-error-in-browser

> , to contact support **where you can open a ticket.**

**Not really**, you can't open a ticket. You can post to the **forum**, a moderator replies saying that a Support Engineer will chime in shortly, which they'll do in **a few days** usually with an unhelpful reply that misses the point entirely, and then the cycle repeats, **every iteration taking longer than the previous one**.

Anyway, **a javascript that throws errors on the client side is a bug**, no matter how bad the configuration is (which, given that I never touched such configuration and left it how it came by default, would in turn be a bug in the default configuration or the installer).

**Comment by php4fan** [comment]
> you are missing key configuration properties. Please review our [installation docs](https://docs.newrelic.com/docs/browser/browser-monitoring/installation/install-browser-monitoring-agent/)

Can you at least point to the specific part of the docs about the supposed configuration properties that are required for the browser script to not throw errors? So at least I could check which one of them may be missing.
What you linked to is the installation instructions, which doesn't mention any of that and only links to the documentation of the ini setting `newrelic.browser_monitoring.auto_instrument`

**Comment by patrickhousley** [comment]
@php4fan I have reached out to our support engineers internally to bring this and the support thread you linked to their attention. I pointed you to the installation instructions because it does not appear as though your installation is correct. You are missing the entire `NREUM.info` block which contains the properties that would replace those `undefined` values you are seeing in the URIs. Our support engineers are better positioned to help you with installation questions and issues.

**Comment by cwli24** [comment]
@php4fan If you're using PHP agent to inject the agent, and you're using their automatic method, the necessary configurations should be handled automatically as well. Are you manually instrumenting and missing the `browser_timing_header/footer` calls?
https://docs.newrelic.com/docs/apm/agents/php-agent/features/browser-monitoring-php-agent/

**Comment by php4fan** [comment]
> missing the browser_timing_header/footer calls?

THANK YOU! Finally some actual insight.

**I was missing the footer call** in some random pages because of a bug of mine.

There's still a bug in the newrelic javascript code in that you don't **check for the existence of NREUM.info** and/or its properties before blindly using possibly undefined values, resulting in requests to the wrong urls, resulting in CORS violation, which is a huge mess. You should add checks and throw your own higher-level exceptions or log messages

**Comment by cwli24** [comment]
We've had problems with `info` being located after the loader code / at the bottom of the body before, and the beacon (url domain) is defaulted. The app ID & license key are not defaulted, but we try sending it anyways. I raised this point and we created a story to short circuit the agent if those are not provided going forward.
(So NREUM.info is being checked again after page load, but we just did not bother considering the "enforcement" of these 2 other configs up until now.)
